src/attached_probe.cpp

```diff
@@ -223,16 +225,142 @@ std::string AttachedProbe::sanitise(const std::string &str)
   return std::regex_replace(str, std::regex("[^A-Za-z0-9_]"), "_");
 }
 
-uint64_t AttachedProbe::offset() const
+struct symbol {
+  std::string name;
+  uint64_t    start;
+  uint64_t    size;
+  uint64_t    address;
+};
+
+static int sym_name_cb(const char *symname, uint64_t start,
+                       uint64_t size, void *p)
 {
-  bcc_symbol sym;
-  int err = bcc_resolve_symname(probe_.path.c_str(), probe_.attach_point.c_str(),
-      probe_.loc, 0, nullptr, &sym);
+  struct symbol *sym = static_cast<struct symbol*>(p);
 
-  if (err)
-    throw std::runtime_error("Could not resolve symbol: " + probe_.path + ":" + probe_.attach_point);
+  if (sym->name == symname)
+  {
+    sym->start = start;
+    sym->size  = size;
+    return -1;
+  }
+
+  return 0;
+}
+
+static int sym_address_cb(const char *symname, uint64_t start,
+                          uint64_t size, void *p)
+{
+  struct symbol *sym = static_cast<struct symbol*>(p);
+
+  if (sym->address >= start && sym->address < (start + size))
```

Comment:
Why not `sym->address == start` instead? If we do `>=`, then we have to worry about nested symbols: https://github.com/iovisor/bcc/blob/master/src/cc/bcc_syms.cc#L380-L399

Reply:
will check, thanks

Reply:
we are checking in here if the provided address fits within the symbol,
not just the symbol start

Reply:
Hmm, I'm still worried about nested symbols. Suppose you have the follow .text layout

```
 0x0              0x8       0x10  
+------------------------------------------------+
| <   symbol A   >|<symbol B>| <rest of symbol A>| 
+------------------------------------------------+
```
where `<symbol B>` is some kind of nested function.

And the callback is run in the following order:
```
sym_address_cb(<symbol A>, 0x0, 24)
sym_address_cb(<symbol B>, 0x8, 8)
...
```

And you were trying to resolve addr 0x9. The real symbol should be `<symbol B>` but you resolve it to `<symbol A>`.

This is my understand, at least. Does this sound right?

Reply:
Hmm, I'm still worried about nested symbols. Suppose you have the follow .text layout

```
 0x0              0x8       0x10  
+------------------------------------------------+
| <   symbol A   >|<symbol B>| <rest of symbol A>| 
+------------------------------------------------+
```
where `<symbol B>` is some kind of nested function.

And the callback is run in the following order:
```
sym_address_cb(<symbol A>, 0x0, 24)
sym_address_cb(<symbol B>, 0x8, 8)
...
```

And you were trying to resolve addr 0x9. The real symbol should be `<symbol B>` but you resolve it to `<symbol A>`.

This is my understand, at least. Does this sound right?

Reply:
not sure there's a good answer for this.. if symbol A spans from 0 to 24
then the correct symbol is also symbol A

in perf we fix overlapping symbols by force ;-) so I've never encounter this problem, nor have I an idea how common this is.. bcc code seems to lookup only based on symbol start address

Reply:
>  if symbol A spans from 0 to 24 then the correct symbol is also symbol A

Good point.

Seems like this symbol name is only for diagnostics. So not super critical. I think we can just emulate perf here :).